pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,43 +88,84 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
+// SaveAlertInstances saves all the provided alert instances to the store. It
+// writes transactions with a maximum count of 400 rows. Larger writes are
+// split into multiple transactions.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	// SQLite has a limit of 999 variables per write. We use 9 per row, so we split up our writes into separate upsert statements.
+	err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		keysPerRow := 9
+		maxRows := 20
+
+		// Prepare the statement to bind each argument to.
+		bigUpsertSQL, err := st.SQLStore.Dialect.UpsertMultipleSQL(
+			"alert_instance",
+			[]string{"rule_org_id", "rule_uid", "labels_hash"},
+			[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"},
+			maxRows)
 		if err != nil {
 			return err
 		}
 
-		alertInstance := &models.AlertInstance{
-			RuleOrgID:         cmd.RuleOrgID,
-			RuleUID:           cmd.RuleUID,
-			Labels:            cmd.Labels,
-			LabelsHash:        labelsHash,
-			CurrentState:      cmd.State,
-			CurrentReason:     cmd.StateReason,
-			CurrentStateSince: cmd.CurrentStateSince,
-			CurrentStateEnd:   cmd.CurrentStateEnd,
-			LastEvalTime:      cmd.LastEvalTime,
+		bigStmt, err := sess.DB().Prepare(bigUpsertSQL)
+		if err != nil {
+			return err
 		}
 
-		if err := models.ValidateAlertInstance(alertInstance); err != nil {
-			return err
+		maxArgs := maxRows * keysPerRow
+		args := make([]interface{}, 0, maxArgs)
+		for _, alertInstance := range cmd {
+			if len(args) >= maxArgs {
+				_, err = bigStmt.ExecContext(ctx, args...)
+				if err != nil {
+					return err
+				}
+				args = args[:0]
+			}
+
+			labelTupleJSON, err := alertInstance.Labels.StringKey()
+			if err != nil {
+				return err
+			}
+
+			if err := models.ValidateAlertInstance(alertInstance); err != nil {
+				return err
+			}
+
+			args = append(args,
+				alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash,
+				alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(),
+				alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
 		}
 
-		params := append(make([]interface{}, 0), alertInstance.RuleOrgID, alertInstance.RuleUID, labelTupleJSON, alertInstance.LabelsHash, alertInstance.CurrentState, alertInstance.CurrentReason, alertInstance.CurrentStateSince.Unix(), alertInstance.CurrentStateEnd.Unix(), alertInstance.LastEvalTime.Unix())
+		if len(args) >= 0 {
```

Comment:
What about this, or, having a separate condition should `len(args)` be `> 0` but not equal to `keysPerRow`?

```suggestion
		if len(args) == keysPerRow {
```